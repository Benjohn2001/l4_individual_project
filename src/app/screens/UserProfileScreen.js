import React, { useEffect, useState } from 'react';
import { Feather } from '@expo/vector-icons';
import { View, Image, TextInput, Text, TouchableOpacity, SafeAreaView, Alert, ActivityIndicator } from 'react-native';
import AppButtonPurple from '../components/AppButtonPurple'
import AppButtonLight from '../components/AppButtonLight';
import TwoButtonsSide from '../components/TwoButtonsSide';
import { COLOURS } from '../assets/colours';
import ProfileBox from '../components/ProfileBox';
import RowItem from '../components/RowItem';
import PillButton from '../components/PillButton';
import { auth } from '../../firebase';
import { getDatabase, ref, set, onValue, orderByChild, query, equalTo, get, child, update, push, remove } from "firebase/database";
import { useRef } from 'react';
import Modal from "react-native-modal";
import TwoButtonStack from '../components/TwoButtonStack';
import * as ImagePicker from 'expo-image-picker';
import { getStorage, getDownloadURL, uploadBytesResumable, uploadBytes, uploadString } from "firebase/storage";
import {ref as stref} from "firebase/storage";
import PressableIcon from '../components/PressableIcon';


const UserProfileScreen = ({ route, navigation }) => {
    
    const [nameFull, setNameFull] = useState("")
    const [userName, setUserName] = useState("")
    const [bio, setBio] = useState("")
    const [pic, setPic] = useState("")
    const [friend, setFriend] = useState(false)
    const db=getDatabase()
    const uid=auth.currentUser.uid

    const user = route.params.user

    useEffect(()=>{
        setNameFull(user.val()["firstName"]+" "+user.val()["lastName"])
        setUserName(user.val()["userName"])
        setBio(user.val()["bio"])
        if(user.val()["profilePic"]!== undefined){
            getDownloadURL(stref(getStorage(), user.val()["profilePic"])).then((url)=>{
                setPic(url)
            })
        }else{
            setPic(Image.resolveAssetSource(require('../assets/defaultProfilePic.png')).uri)
        }
        const checkFriend = async () =>{
            const friendsRef = ref(getDatabase(), "/friends/"+uid)
            const dataFr = await get(friendsRef)
            if(dataFr.val() !== null){
                dataFr.forEach(c => {
                    if(c.val()["user"]==user.key){
                        setFriend(true)
                    }
                })
            }
        }

        checkFriend()
        
    },[])
    
    return (
        <SafeAreaView className="flex-1 bg-primaryPurple" >
            <View className="flex-row justify-between pt-14 mx-10">
                <PressableIcon
                        onPress={() => {
                            navigation.goBack()
                        }}
                        icon="arrow-left"
                        size={40}
                        color="black"
                    />
                <Text className="font-bold ml-auto mr-auto text-3xl">
                    User Profile
                </Text>
            </View>
            {pic ? <View className="items-center justify-center pt-14">
                <Image 
                    className="w-44 h-44 rounded-full" 
                    source = {{uri: pic}}
                />
                <View className="py-12">
                    <ProfileBox
                        name={nameFull}
                        username={userName}
                        bio={bio}
                    />
                </View>
                {friend ? 
                    <AppButtonPurple
                        title={"Remove friend"}
                        onPress={async ()=>{
                            const dref=query(ref(db, "/friends/"+uid))
                            const data = await get(dref)
                            if(data.val() !== null){
                                data.forEach(c => {

                                    if(c.val()["user"]==user.key){
                                        remove(ref(db, "/friends/"+uid+"/"+c.key))
                                    }
                                })
                            }
                            Alert.alert("Success","Friend removed succesfully")
                            setFriend(false)
                            }
                        }
                    />
                :
                <AppButtonPurple
                    title={"Add friend"}
                    onPress={async ()=>{
                        const pref=push(ref(db, '/friends/' + uid ))
                        await set(pref,{
                            user: user.key
                        })
                        Alert.alert("Success","Friend added succesfully")
                        setFriend(true)
                    }}
                />
                }
            </View> : 
            <View className="justify-center items-center flex-1">
                <ActivityIndicator size="large" color="#6B4EFF"  />
                <Text className="text-center">Loading Profile</Text>
            </View>
            }
        </SafeAreaView>
    );
};

export default UserProfileScreen;