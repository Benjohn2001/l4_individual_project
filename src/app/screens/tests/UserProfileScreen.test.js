import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react-native";
import UserProfileScreen from "../UserProfileScreen";


jest.mock("firebase/database", () => ({
  getDatabase: jest.fn().mockReturnThis(),
  push: jest.fn().mockReturnThis(),
  ref: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
  get: jest.fn().mockReturnThis(),
  query: jest.fn().mockReturnThis(),
  onValue: jest.fn().mockReturnThis(),
  orderByChild: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/storage", () => ({
  getStorage: jest.fn().mockReturnThis(),
}));

const navigation = {
  navigate: jest.fn(),
  push: jest.fn(),
  goBack: jest.fn(),
};

const route = {
    params: {
      user: {
        val: jest.fn().mockReturnThis(),
      },
    },
  };

jest.mock("firebase/app", () => ({
  auth: jest.fn().mockReturnThis(),
  initializeApp: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/auth", () => ({
  auth: jest.fn().mockReturnThis(),
  getAuth: () => ({
    currentUser: jest.fn().mockReturnThis(),
    signOut: () => ({
      then: () => ({
        catch: jest.fn(),
      }),
    }),
  }),
}));

jest.mock("@react-navigation/native", () => ({
    useIsFocused: jest.fn(),
  }));

test("Check UserProfileScreen working correctly", () => {
  render(<UserProfileScreen navigation={navigation} route={route} />);
  fireEvent.press(screen.getByTestId("pressableIcon"))
  expect(navigation.goBack).toHaveBeenCalledTimes(1)
  const domTree = renderer
    .create(<UserProfileScreen route={route} />)
    .toJSON();
  expect(domTree).toMatchSnapshot();
});
