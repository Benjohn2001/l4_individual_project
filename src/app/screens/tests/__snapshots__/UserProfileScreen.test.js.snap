// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check UserProfileScreen working correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "backgroundColor": "#C6C4FF",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "justifyContent": "space-between",
        },
        Object {
          "paddingTop": 56,
        },
        Object {
          "marginLeft": 40,
          "marginRight": 40,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="pressableIcon"
      >
        <Text />
      </View>
    </View>
    <Text
      style={
        Array [
          Object {
            "fontWeight": "700",
          },
          Object {
            "marginLeft": "auto",
          },
          Object {
            "marginRight": "auto",
          },
          Object {
            "fontSize": 30,
            "lineHeight": 36,
          },
        ]
      }
    >
      User Profile
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
        },
        Object {
          "alignItems": "center",
        },
        Object {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <ActivityIndicator
      color="#6B4EFF"
      size="large"
    />
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      Loading Profile
    </Text>
  </View>
</RCTSafeAreaView>
`;
